File: Quasics2023Code/ChargedUp/src/main/cpp/commands/ArcadeDrive.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "commands/ArcadeDrive.h"

ArcadeDrive::ArcadeDrive(Drivebase* drivebase,
                         std::function<double()> powerFunction,
                         std::function<double()> turnFunction)
    : m_drivebase(drivebase),
      m_powerFunction(powerFunction),
      m_turnFunction(turnFunction) {
  // Use addRequirements() here to declare subsystem dependencies.
  AddRequirements(drivebase);
}

// Called when the command is initially scheduled.
void ArcadeDrive::Initialize() {
  UpdateSpeeds();
}

// Called repeatedly when this Command is scheduled to run
void ArcadeDrive::Execute() {
  UpdateSpeeds();
}

// Called once the command ends or is interrupted.
void ArcadeDrive::End(bool interrupted) {
  m_drivebase->Stop();
}

void ArcadeDrive::UpdateSpeeds() {
  m_drivebase->ArcadeDrive(m_powerFunction(), m_turnFunction());
}

File: Quasics2025Code/ReefScape 2025/src/main/java/frc/robot/commands/ArcadeDrive.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.drivebase.IDrivebase;
import frc.robot.subsystems.drivebase.RealDrivebase;

import java.util.function.Supplier;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.LinearVelocity;

/* You should consider using the more terse Command factories API instead https://docs.wpilib.org/en/stable/docs/software/commandbased/organizing-command-based.html#defining-commands */
public class ArcadeDrive extends Command {
  private final IDrivebase m_drivebase;
  private final Supplier<Double> m_speedSupplier;
  private final Supplier<Double> m_rotationSupplier;

  /** Creates a new ArcadeDrive. */
  public ArcadeDrive(
    IDrivebase drivebase, Supplier<Double> leftSupplier, Supplier<Double> rightSupplier) {
      m_speedSupplier = leftSupplier;
      m_rotationSupplier = rightSupplier;
      m_drivebase = drivebase;

      addRequirements(drivebase);
    // Use addRequirements() here to declare subsystem dependencies.
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {
  updateSpeeds();
  }

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    updateSpeeds();
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {
    m_drivebase.stop();
  }

  private void updateSpeeds() {
    final double leftInput = m_speedSupplier.get();
    final double rightInput = m_rotationSupplier.get();

    LinearVelocity forwardSpeed = IDrivebase.MAX_SPEED.times(leftInput);
    AngularVelocity rotationSpeed = IDrivebase.MAX_ANGULAR_SPEED.times(rightInput);

    m_drivebase.arcadeDrive(forwardSpeed, rotationSpeed);
  }
}
